### PR TITLE
feat: shared processing in plot callbacks

### PR DESCRIPTION
**Pull request by VeraChristina**
## Description
Reduce the duplication of post-processing in the plotting callbacks by introducing a plot playload that all callbacks fetch from the plot adapters and that caches post-processed tensors on first access.

## What issue or task does this change relate to?
https://github.com/ecmwf/anemoi-core/issues/1135

##  Additional notes ##

***As a contributor to the Anemoi framework, please ensure that your changes include unit tests, updates to any affected dependencies and documentation, and have been tested in a parallel setting  (i.e., with multiple GPUs). As a reviewer, you are also responsible for verifying these aspects and requesting changes if they are not adequately addressed. For guidelines about those please refer to https://anemoi.readthedocs.io/en/latest/***

By opening this pull request, I affirm that all authors agree to the [Contributor License Agreement.](https://github.com/ecmwf/codex/blob/main/Legal/contributor_license_agreement.md)

### CODE REVIEW

File: training/src/anemoi/training/diagnostics/callbacks/plot.py [hunk 1]
```diff
@@ -903,22 +886,54 @@ def process(
             self.latlons = {}
 
         if dataset_name not in self.latlons:
-            self.latlons[dataset_name] = pl_module.model.model._graph_data[dataset_name].x.detach()
-            self.latlons[dataset_name] = np.rad2deg(self.latlons[dataset_name].cpu().numpy())
+            # Fallback: extract latlons if not already populated by prepare_payload
+            self.latlons[dataset_name] = np.rad2deg(
```
Comment (anaprietonem) [inline training/src/anemoi/training/diagnostics/callbacks/plot.py]: why is this needed? shouldn't we check if that's on the payload? * okey I do see above that self.latlons is actually coming from the payload but still not clear when the Fallback is hit

File: training/src/anemoi/training/diagnostics/callbacks/plot.py [hunk 2]
```diff
@@ -903,22 +886,54 @@ def process(
...
+                tuple(
+                    pl_module.plot_adapter.select_members(
+                        denormed_output[step_idx, self.sample_idx : self.sample_idx + 1],
+                        members,
+                    )
+                    for step_idx in range(denormed_output.shape[0])
+                ),
+            )
+        else:
+            # Fallback for callers without a payload (e.g. direct process() calls)
```
Comment (anaprietonem) [inline training/src/anemoi/training/diagnostics/callbacks/plot.py]: similarly when is this Fallback hit?

Comment (anaprietonem) [reply training/src/anemoi/training/diagnostics/callbacks/plot.py]: If we do need the Fallback, it would good  to refactor the below to some utility function to avoid the hardcoding of very similar code between the payload computation and the fallback

File: training/src/anemoi/training/diagnostics/callbacks/plot.py [hunk 3]
```diff
@@ -903,22 +886,54 @@ def process(
...
+                        self.post_processors[dataset_name](x[dataset_name][:, ...].detach().cpu(), in_place=False)[
+                            self.sample_idx : self.sample_idx + 1
+                        ],
+                        members,
+                    )
+                    for x in outputs.predictions
+                ),
+            )
 
+        output_tensor = pl_module.plot_adapter.prepare_plot_output_tensor(output_tensor)
```
Comment (anaprietonem) [inline training/src/anemoi/training/diagnostics/callbacks/plot.py]: this is equivalent to the function below for process_output_tensor so why do we need that function still? can't we get rid of it?

**Review by anaprietonem — COMMENTED** [review]
Nice work Vera! left a few comments to clarify

File: training/src/anemoi/training/diagnostics/callbacks/plot.py (hunk 1 above)
Comment (VeraChristina) [reply training/src/anemoi/training/diagnostics/callbacks/plot.py]: good point, it's only hit when prepare_payload wasn't called before process (which doesn't happen in the current use of the class). I'll remove the fallback and make this clear in the docstring

File: training/src/anemoi/training/diagnostics/callbacks/plot.py (hunk 2 above)
Comment (VeraChristina) [reply training/src/anemoi/training/diagnostics/callbacks/plot.py]: as above

File: training/src/anemoi/training/diagnostics/callbacks/plot.py (hunk 3 above)
Comment (VeraChristina) [reply training/src/anemoi/training/diagnostics/callbacks/plot.py]: the method process_output_tensor was added in the recent merge of main. As far as I understand it's used to process auxiliary outputs (from plot_kwargs) not present in the payload. There is some overlap with the code here, but process_output_tensor denormalises outputs, while here we use already denormalised outputs -- I could extract the common part into a helper function but that's only three lines of code.

File: training/src/anemoi/training/diagnostics/callbacks/plot.py [hunk 4]
```diff
@@ -365,47 +365,26 @@ def on_validation_batch_end(
...
+            self.post_processors = payload.post_processors
+            self.latlons = payload.latlons
+            self._payload = payload
 
             plot_kwargs = self._plot_kwargs_from_output(pl_module, output)
             output = TrainingStepOutput(
                 loss=output.loss,
                 metrics=output.metrics,
-                predictions=gathered_predictions,
+                predictions=payload.outputs[1],
```
Comment (anaprietonem) [inline training/src/anemoi/training/diagnostics/callbacks/plot.py]: could this be cleaner if we do something like?

```

@dataclass
class GatheredOutputs:
    loss: torch.Tensor
    predictions: list[dict[str, torch.Tensor]]

gathered = GatheredOutputs(
    loss=output.loss,
    predictions=[
        {
            dataset_name: pl_module.allgather_batch(dataset_pred, dataset_name)
            for dataset_name, dataset_pred in pred.items()
        }
        for pred in preds
    ],
)

self._cached_payload = PlotPayload(
    batch_idx=batch_idx,
    batch=gathered_batch,
    outputs=gathered,  # ← now a named object
    post_processors=post_processors,
    latlons=latlons,
    feature_indices=feature_indices,
)
```

Though looking at this a bit more, I am getting confused why we have the loss in the output payload payload.outputs[0] but in TrainingStepOutput we use output.loss?

File: training/src/anemoi/training/diagnostics/callbacks/plot.py (hunk 3 above)
Comment (anaprietonem) [reply training/src/anemoi/training/diagnostics/callbacks/plot.py]: Yes I want to check better the need for process_output_tensor but I think that's out of the scope for this PR, so a helper function will be fine.

File: training/src/anemoi/training/diagnostics/callbacks/plot.py (hunk 4 above)
Comment (VeraChristina) [reply training/src/anemoi/training/diagnostics/callbacks/plot.py]: good catch, it's a reference to the legacy tuple format of outputs (which was replaced with TrainingStepOutput), I had missed this after merging, have removed now

File: training/src/anemoi/training/diagnostics/callbacks/plot.py (hunk 3 above)
Comment (VeraChristina) [reply training/src/anemoi/training/diagnostics/callbacks/plot.py]: done!

**Review by anaprietonem — DISMISSED** [review]
Thanks for addressing the comments Vera, nice work! LGTM to be merged

**Comment by anaprietonem** [comment]
Closing the PR, due to the issues with memory management. Feature implemented under https://github.com/ecmwf/anemoi-core/commit/2a72fd81e9d78351100ed3aa783f427308d5d68c
